pkg/dns/dns.go

```diff
@@ -0,0 +1,417 @@
+}
+
+func (kd *KubeDNS) generateRecordsForHeadlessService(e *kapi.Endpoints, svc *kapi.Service) error {
+	// TODO: remove this after v1.4 is released and the old annotations are EOL
+	podHostnames, err := getPodHostnamesFromAnnotation(e.Annotations)
+	if err != nil {
+		return err
+	}
+	subCache := NewTreeCache()
+	glog.V(4).Infof("Endpoints Annotations: %v", e.Annotations)
+	for idx := range e.Subsets {
+		for subIdx := range e.Subsets[idx].Addresses {
+			address := &e.Subsets[idx].Addresses[subIdx]
+			endpointIP := address.IP
+			recordValue, endpointName := getSkyMsg(endpointIP, 0)
+			if hostLabel, exists := getHostname(address, podHostnames); exists {
+				endpointName = hostLabel
+			}
+			subCache.SetEntry(endpointName, recordValue)
+			for portIdx := range e.Subsets[idx].Ports {
+				endpointPort := &e.Subsets[idx].Ports[portIdx]
+				if endpointPort.Name != "" && endpointPort.Protocol != "" {
+					srvValue := kd.generateSRVRecordValue(svc, int(endpointPort.Port), endpointName)
+					subCache.SetEntry(endpointName, srvValue, "_"+strings.ToLower(string(endpointPort.Protocol)), "_"+endpointPort.Name)
+				}
+			}
+		}
+	}
+	subCachePath := append(kd.domainPath, serviceSubdomain, svc.Namespace)
+	kd.cache.SetSubCache(svc.Name, subCache, subCachePath...)
+	return nil
+}
+
+func getHostname(address *kapi.EndpointAddress, podHostnames map[string]endpoints.HostRecord) (string, bool) {
+	if len(address.Hostname) > 0 {
+		return address.Hostname, true
+	}
+	if hostRecord, exists := podHostnames[address.IP]; exists && validation.IsDNS1123Label(hostRecord.HostName) {
+		return hostRecord.HostName, true
+	}
+	return "", false
+}
+
+func getPodHostnamesFromAnnotation(annotations map[string]string) (map[string]endpoints.HostRecord, error) {
+	hostnames := map[string]endpoints.HostRecord{}
+
+	if annotations != nil {
+		if serializedHostnames, exists := annotations[endpoints.PodHostnamesAnnotation]; exists && len(serializedHostnames) > 0 {
+			err := json.Unmarshal([]byte(serializedHostnames), &hostnames)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return hostnames, nil
+}
+
+func (kd *KubeDNS) generateSRVRecordValue(svc *kapi.Service, portNumber int, cNameLabels ...string) *skymsg.Service {
+	cName := strings.Join([]string{svc.Name, svc.Namespace, serviceSubdomain, kd.domain}, ".")
+	for _, cNameLabel := range cNameLabels {
+		cName = cNameLabel + "." + cName
+	}
+	recordValue, _ := getSkyMsg(cName, portNumber)
+	return recordValue
+}
+
+// Generates skydns records for a headless service.
+func (kd *KubeDNS) newHeadlessService(service *kapi.Service) error {
+	// Create an A record for every pod in the service.
+	// This record must be periodically updated.
+	// Format is as follows:
+	// For a service x, with pods a and b create DNS records,
+	// a.x.ns.domain. and, b.x.ns.domain.
+	key, err := kcache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		return err
+	}
+	e, exists, err := kd.endpointsStore.GetByKey(key)
```

Comment:
Same comment as above, but for using a service key in endpoints lookup. Please consider adding a comment.